File: src/main/java/com/BoardingHouse/BoardingHouse_Management_System/service/auth/AuthenticationService.java
package com.BoardingHouse.BoardingHouse_Management_System.service.auth;

import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

@Service
public class AuthenticationService {

    private final AuthenticationManager authenticationManager;
    private final PasswordEncoder passwordEncoder;

    public AuthenticationService(AuthenticationManager authenticationManager,
                                 PasswordEncoder passwordEncoder) {
        this.authenticationManager = authenticationManager;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Authenticate a username/password using Spring Security.
     * Throws AuthenticationException if invalid.
     */
    public Authentication authenticate(String username, String password)
            throws AuthenticationException {

        if (!StringUtils.hasText(username) || !StringUtils.hasText(password)) {
            throw new IllegalArgumentException("Username and password must not be empty");
        }

        UsernamePasswordAuthenticationToken token =
                new UsernamePasswordAuthenticationToken(username, password);

        Authentication authentication = authenticationManager.authenticate(token);

        // Store authentication in security context (important!)
        SecurityContextHolder.getContext().setAuthentication(authentication);

        return authentication;
    }

    /** Hash password before saving to DB */
    public String encodePassword(String rawPassword) {
        if (!StringUtils.hasText(rawPassword)) {
            throw new IllegalArgumentException("Password cannot be empty");
        }
        return passwordEncoder.encode(rawPassword);
    }

    /** Check raw vs hashed password */
    public boolean matchesPassword(String rawPassword, String encodedPassword) {
        return passwordEncoder.matches(rawPassword, encodedPassword);
    }

    /** Logout current user */
    public void clearAuthentication() {
        SecurityContextHolder.clearContext();
    }
}

File: services/authentication_service.py
from fastapi import HTTPException, status

from config.security_config import password_encoder
from config.jwt_config import jwt_config


class AuthenticationService:
    

    def authenticate(self, username: str, password: str, hashed_password_from_db: str) -> dict:
       
        if not username or not password:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username and password must not be empty",
            )

        # Verify password - equivalent to DaoAuthenticationProvider doing the same internally
        if not password_encoder.matches(password, hashed_password_from_db):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials",
            )

        # Issue tokens - replaces storing auth in SecurityContextHolder for stateless JWT flow
        access_token = jwt_config.generate_token(username)
        refresh_token = jwt_config.generate_refresh_token(username)

        return {
            "username": username,
            "access_token": access_token,
            "refresh_token": refresh_token,
            "token_type": "Bearer",
        }

    def encode_password(self, raw_password: str) -> str:
        """
       
        Call this before persisting a new user.
        """
        if not raw_password:
            raise ValueError("Password cannot be empty")
        return password_encoder.encode(raw_password)

    def matches_password(self, raw_password: str, encoded_password: str) -> bool:
       
        return password_encoder.matches(raw_password, encoded_password)

    def clear_authentication(self, request) -> None:
       
        request.state.username = None
        request.state.authenticated = False


# Singleton service instance (equivalent to Spring's @Service singleton bean)
authentication_service = AuthenticationService()